File: dojoagents/dashboard/web/src/types/agent.ts
export interface AgentModelItem {
  id: string;
  label: string;
  provider: string;
  model: string;
  available: boolean;
  unavailable_reason?: string | null;
}

export interface AgentModelsResponse {
  default_model_id: string;
  gemini_configured: boolean;
  zhipu_configured: boolean;
  agent_ready: boolean;
  models: AgentModelItem[];
}

export type AgentChatRole = 'user' | 'assistant';

export type AgentApiContentPart =
  | { type: 'text'; text: string }
  | { type: 'image_url'; image_url: { url: string } };

export type AgentApiMessage = {
  role: AgentChatRole;
  content: string | AgentApiContentPart[];
};

export type AgentToolActivityStatus = 'running' | 'done' | 'error';

export interface AgentToolActivityItem {
  callId?: string;
  tool: string;
  status: AgentToolActivityStatus;
  latencyMs?: number;
  error?: string | null;
  arguments?: Record<string, unknown>;
  data?: Record<string, unknown> | null;
  resultSummary?: string | null;
  resultContent?: string | null;
  vizBlocks?: import('./agentViz').AgentVizBlock[];
}

export interface AgentEvalHintItem {
  id: string;
  issues: string[];
}

export interface AgentThinkBlock {
  id: string;
  text: string;
  collapsed: boolean;
  done: boolean;
}

export type AgentActivityStep =
  | { kind: 'think'; id: string; block: AgentThinkBlock }
  | { kind: 'tool'; id: string; item: AgentToolActivityItem }
  | { kind: 'eval'; id: string; hint: AgentEvalHintItem };

export interface AgentChatImageAttachment {
  dataUrl: string;
  name?: string;
}

export interface AgentChatMessage {
  role: AgentChatRole;
  content: string;
  /** Pasted or uploaded images attached to a user message. */
  images?: AgentChatImageAttachment[];
  /** Chronological stream of thinking, tools, and eval hints. */
  activitySteps?: AgentActivityStep[];
  /** @deprecated Migrated into activitySteps for display order. */
  toolActivity?: AgentToolActivityItem[];
  /** @deprecated Migrated into activitySteps for display order. */
  thinkBlocks?: AgentThinkBlock[];
  /** @deprecated Migrated into activitySteps for display order. */
  evalHints?: AgentEvalHintItem[];
}

export type AgentLocale = 'zh' | 'en';

export interface AgentChatRequest {
  session_id: string;
  model_id: string;
  messages: AgentApiMessage[];
  locale?: AgentLocale;
  dashboard_tab?: string;
  use_tools?: boolean;
  max_tool_steps?: number;
  exclude_mutating_tools?: boolean;
}

export interface AgentToolTraceItem {
  call_id?: string;
  tool: string;
  arguments: Record<string, unknown>;
  ok: boolean;
  latency_ms: number;
  truncated: boolean;
  error?: string | null;
  data?: Record<string, unknown> | null;
  viz_blocks?: import('./agentViz').AgentVizBlock[];
  artifacts?: Record<string, unknown>[];
  resource_changes?: Record<string, unknown>[];
}

export interface AgentChatResponse {
  model_id: string;
  message: AgentChatMessage;
  tool_trace?: AgentToolTraceItem[];
  tool_steps?: number;
}

export interface AgentSession {
  id: string;
  title: string;
  modelId: string;
  messages: AgentChatMessage[];
  createdAt: number;
  updatedAt: number;
  /** Monotonic local version used to resolve async persistence races. */
  revision?: number;
}

export interface AgentSessionStore {
  activeSessionId: string | null;
  sessions: AgentSession[];
}

export type AgentStreamEvent =
  | { type: 'delta'; text: string }
  | { type: 'think_start' }
  | { type: 'think_delta'; text: string }
  | { type: 'think_end' }
  | { type: 'phase'; phase: 'planning' | 'tools' | 'answering' }
  | { type: 'retry'; attempt: number; max_attempts: number; text: string }
  | { type: 'tool_start'; call_id?: string; tool: string; arguments: Record<string, unknown> }
  | {
      type: 'tool_result';
      call_id?: string;
      tool: string;
      ok: boolean;
      content?: string;
      latency_ms: number;
      truncated: boolean;
      error?: string | null;
      data?: Record<string, unknown> | null;
      viz_blocks?: import('./agentViz').AgentVizBlock[];
      resource_changes?: Record<string, unknown>[];
    }
  | { type: 'eval_hint'; text: string; issues: string[] }
  | { type: 'done'; model_id: string; tool_trace?: AgentToolTraceItem[]; tool_steps?: number }
  | { type: 'error'; message: string };

File: dojoagents/dashboard/schemas/agent.py
from __future__ import annotations

from typing import List, Literal

from pydantic import BaseModel, Field


class AgentModelItem(BaseModel):
    id: str = Field(..., description="Stable model identifier for API requests")
    label: str = Field(..., description="Display name, e.g. Gemini-3.5")
    provider: str = Field(..., description="Provider slug: google, openai, anthropic, …")
    model: str = Field("", description="Concrete provider model id")
    available: bool = Field(..., description="Whether the model can be selected and used")
    unavailable_reason: str | None = Field(None, description="Reason the model cannot be selected")


class AgentModelsResponse(BaseModel):
    default_model_id: str = Field(..., description="Default selected model id")
    gemini_configured: bool = Field(
        False,
        description="True when GEMINI_API_KEY is set on the server",
    )
    zhipu_configured: bool = Field(
        False,
        description="True when a Zhipu/GLM provider is configured on the server",
    )
    agent_ready: bool = Field(
        False,
        description="True when at least one model is available",
    )
    models: List[AgentModelItem] = Field(default_factory=list)


class AgentChatMessage(BaseModel):
    role: Literal["user", "assistant"] = Field(..., description="Conversation role")
    content: str = Field(..., min_length=1, description="Plain-text message body")


class AgentChatRequest(BaseModel):
    model_id: str = Field(..., min_length=1, description="Model id from /agent/models")
    messages: List[AgentChatMessage] = Field(..., min_length=1, max_length=50)


class AgentChatResponse(BaseModel):
    model_id: str
    message: AgentChatMessage
